lib/Dialect/ESI/ESIServices.cpp

```diff
@@ -15,24 +15,131 @@
 #include "circt/Dialect/HW/HWOps.h"
 #include "circt/Dialect/MSFT/MSFTPasses.h"
 
+#include "mlir/IR/BlockAndValueMapping.h"
 #include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
 
 #include <memory>
 
 using namespace circt;
 using namespace circt::esi;
 
+static DenseMap<Attribute, ServiceGeneratorDispatcher::ServiceGeneratorFunc>
+DefaultTableCreator(MLIRContext *);
+
+ServiceGeneratorDispatcher ServiceGeneratorDispatcher::defaultDispatcher() {
+  return ServiceGeneratorDispatcher(DefaultTableCreator, false);
+}
+
+LogicalResult ServiceGeneratorDispatcher::generate(ServiceImplementReqOp req) {
+  // Create the lookup table lazily.
+  if (!lookupTable.hasValue())
+    lookupTable = create(req.getContext());
+
+  // Lookup based on 'impl_type' attribute and pass through the generate request
+  // if found.
+  auto genF = lookupTable->find(req.impl_typeAttr());
+  if (genF == lookupTable->end()) {
+    if (failIfNotFound)
+      return req.emitOpError("Could not find service generator for attribute '")
+             << req.impl_typeAttr() << "'";
+    return success();
+  }
+  return genF->second(req);
+}
+
+/// The generator for the "cosim" impl_type.
+static LogicalResult instantiateCosimEndpoints(ServiceImplementReqOp req) {
+  auto *ctxt = req.getContext();
+  OpBuilder b(req);
+  Block *portReqs = &req.portReqs().getBlocks().front();
+  Value clk = req.getOperand(0);
+  Value rst = req.getOperand(1);
+
+  // Determine which EndpointID this generator should start with.
+  uint64_t epIdCtr = 1000; // Default EpID counter.
+  if (req.impl_opts()) {
+    auto opts = req.impl_opts()->getValue();
+    for (auto nameAttr : opts) {
+      if (nameAttr.getName().getValue() == "EpID_start") {
```

Comment:
The fact that there is one attribute allowed in this dictionary and everything else is an error makes me wonder why it is a dictionary attr in the first place? Why not an Optional<IntegerAttr> for endpoint ID start? If you don't plan to allow arbitrary things in the dictionary, I don't see the point of a dictionary instead of enumerating them in the ODS for the op.

If you do plan to keep a dictionary, you shouldn't have to loop over it. I believe there is a C++ API where you can ask for the "EpID_start" key directly.

Reply:
This isn't just for Cosim. `impl_opts` is user/implementation defined at compiler-runtime. "EpID_start" is the only option defined by the 'cosim' implementation, so if the "cosim" implementation is used I want to check all of the options for correctness.

Reply:
Ok, I see. There is opts.getNamed if you just want to get EpID_start, but I guess you intend to verify any other attributes are not present?

Reply:
Yeah, I've found it's useful in these sorts of semi-structured / free-form dicts to check that there are no unexpected fields. Helps mitigate typos.